File: notebooks/korea_aerospace_university/Functions/sdol_tsa.m
function [ta,t] = sdol_tsa(x,sr,tach,ppr,Nr)
% =========================================================================
% This code is programmed by System Design Optimization Lab (SDOL) at Korea
% Aerospace University (KAU)
% ============================== Input ====================================
% x: Vibration signal
% sr: Sampling frequency
% tach: Tachometer signal
% ppr: The number of pulses per 1 revolution
% Nr: The number of rotation to calculate TSA
% ============================== Output ===================================
% ta: Time synchronous average of signal x
% t: Sample time correspond to ta
% =========================================================================
if nargin == 4
    Nr = 1;
end
x = x(:); t = (0:1/sr:(length(x)-1)/sr)';
ppr = ppr*Nr;
if rem(length(tach),ppr)
    nrev = floor(length(tach)/ppr);
else
    nrev = length(tach)/ppr-1;
end

T = nan(nrev,1);
for j = 1:nrev
    T(j) = tach(ppr*j+1);
end
T = [tach(1); T]; T(T>t(end)) = [];

nrev = length(T)-1; N = round(min(diff(T),[],1)*sr); 
resamp = nan(nrev,N);
for j = 2:nrev+1
    tmp = (T(j)-T(j-1))/N; tmp = tmp*(0:N-1);
    resamp(j-1,:) = T(j-1)+tmp;
end
resx = interp1(t,x,resamp,'linear');
ta = mean(resx)';
f = 1/min(diff(T)); sr_resamp = N*f;
t = (0:N-1)'/sr_resamp;

end

File: notebooks/korea_aerospace_university/Functions/sdol_tsa.py
import numpy as np
from scipy.interpolate import interp1d

# =========================================================================
# This code is programmed by System Design Optimization Lab (SDOL) at Korea
# Aerospace University (KAU)
# ============================== Input ====================================
# x: Vibration signal
# sr: Sampling frequency
# tach: Tachometer signal
# ppr: The number of pulses per 1 revolution
# Nr: The number of rotation to calculate TSA
# ============================== Output ===================================
# ta: Time synchronous average of signal x
# t: Sample time correspond to ta
# =========================================================================

def sdol_tsa(x,sr,tach,ppr,Nr=1):

    t = np.arange(0, len(x) / sr, 1 / sr)
    ppr = ppr*Nr
    if np.remainder(len(tach),ppr) == 0:
        nrev=int(len(tach)/ppr-1)
    else:
        nrev = int(np.floor(len(tach) / ppr))

    T = np.empty(shape=[nrev, 1]) #cycle start time

    for j in range(nrev):
        T[j] = tach[ppr * (j + 1)]

    T = np.insert(T, 0, tach[0])
    ind = np.where(T < t[-1]) #until maximum measurement time
    T = T[ind]

    nrev = len(T) - 1;
    N = round(min(np.diff(T)) * sr)
    resample = np.empty(shape=[nrev, N])
    for j in range(1, nrev + 1):
        tmp = (T[j] - T[j - 1]) / N; tmp = tmp * range(N)
        resample[j - 1, :] = T[j - 1] + tmp

    x = np.ravel(x, order='C')
    resx = interp1d(t, x)(resample)
    ta = np.mean(resx, axis=0)
    ta = np.reshape(ta,(len(ta),1))
    f = 1 / min(np.diff(T))
    sr_resamp = N * f
    t = range(N) / sr_resamp
    
    return ta, t

